### PR TITLE
feat(dataframe): support output to a single CSV file

### PR DESCRIPTION
**Pull request by hongzmsft**
The method allows save a dask data frame to a single CSV file, which
works well with other data processing tools and libraries.

Since specifying a full path to the existing `to_csv` method leads to
the creation of a folder of the given name, this behaviour cannot be
implemented within the existing `to_csv` method without breaking its
existing behaviour. Thus, a new method is provided instead.

- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask`

Closes https://github.com/dask/distributed/issues/2982

### CODE REVIEW

**Comment by TomAugspurger** [comment]
Dose `df.compute().to_csv("path/to/csv")` work instead?

**Comment by hongzmsft** [comment]
@TomAugspurger This is for the data that do not fit into the memory. I believe we should not always use `.compute()` for those cases. This new method provides a way to write the results out in a way that other libraries might be able to read without post-processing or other magics. One case is to compile a huge CSV file for LightGBM to do two-round loading or XGBoost to do out-of-core training.

File: dask/dataframe/io/csv.py [hunk 1]
```diff
@@ -768,7 +770,155 @@ def to_csv(
...
+
+    values = [to_csv_chunk(dfs[0], first_file, **kwargs)]
+
+    if len(dfs) > 1:
+        kwargs["header"] = False
+        values.extend([to_csv_chunk(d, append_file, **kwargs) for d in dfs[1:]])
+
+    if compute:
+        for value in values:
+            delayed(value).compute(scheduler=scheduler)
```
[the remark this answers is not in the document]
Comment (hongzmsft) [reply dask/dataframe/io/csv.py]: @mrocklin I pushed a version as you suggested. I am not quite sure whether it exactly fits your requirement. Critiques are welcomed!

File: dask/dataframe/io/csv.py [hunk 2]
```diff
@@ -768,7 +770,130 @@ def to_csv(
...
+
+    Returns
+    -------
+    The names of the file written if they were computed right away
+    If not, a single delayed task associated to the writing of the
+    files.
+    """
+    if PY2:
+        default_encoding = None
+        default_mode = "wb"
```
Comment (mrocklin) [inline dask/dataframe/io/csv.py]: We no longer support Python 2, so you can safely drop this.

File: dask/dataframe/io/csv.py [hunk 3]
```diff
@@ -768,7 +770,130 @@ def to_csv(
...
+    else:
+        default_encoding = "utf-8"
+        default_mode = "wt"
+    encoding = kwargs.get("encoding", default_encoding)
+    mode = kwargs.get("mode", default_mode)
```
Comment (mrocklin) [inline dask/dataframe/io/csv.py]: And we can probably move these into normal kwargs

**Comment by hongzmsft** [comment]
@mrocklin I have added a second commit to merge `to_csv` and `to_single_csv` as you proposed. The implementation is not necessarily cleaner, but API seems neater.

My understanding is that `dask.dataframe.DataFrame.to_csv` already behaves differently from `pandas.DataFrame.to_csv`. Therefore I expect that if we are to use the keyword argument, `single_file` will be `False` by default.

**Comment by jrbourbeau** [comment]
I'd prefer to add a `single_file=` parameter to `df.to_csv` instead of a new `df.to_single_csv` method. As @hongzmsft mentioned, we already have Dask-specific keyword arguments in `df.to_csv` (e.g. `compute`, `name_function`, `scheduler`, etc.)

**Comment by jrbourbeau** [comment]
It looks like tests are passing, but there are some code formatting issues causing the CI to fail. @hongzmsft running `flake8` and `black` locally should fix this (https://docs.dask.org/en/latest/develop.html#code-formatting)

**Comment by hongzmsft** [comment]
@jrbourbeau Fixed.

File: dask/dataframe/io/csv.py [hunk 4]
```diff
@@ -592,14 +592,18 @@ def read(
 read_fwf = make_reader(pd.read_fwf, "read_fwf", "fixed-width")
 
 
-def _write_csv(df, fil, **kwargs):
+def _write_csv(df, fil, result, **kwargs):
```
Comment (mrocklin) [inline dask/dataframe/io/csv.py]: This name is strange in the many csv files case.  I recommend `token` instead, which demotes the importance of this keyword for reviewers.

File: dask/dataframe/io/csv.py [hunk 5]
```diff
@@ -592,14 +592,18 @@ def read(
...
     with fil as f:
         df.to_csv(f, **kwargs)
+    return result
```
Comment (mrocklin) [inline dask/dataframe/io/csv.py]: Do we need to return anything here?  Or is it enough just to return None?

File: dask/dataframe/io/csv.py [hunk 6]
```diff
@@ -609,7 +613,8 @@ def to_csv(
     **kwargs
 ):
     """
-    Store Dask DataFrame to CSV files
+    Store Dask DataFrame to either a single CSV file or a group of CSV
+    files.
```
Comment (mrocklin) [inline dask/dataframe/io/csv.py]: I would prefer not to advertise the single file feature as prominently as this.  It is likely to perform much worse.  Instead, I would prefer that we keep the feature limited to just the keyword description below, and revert this change to the docstring title.

File: dask/dataframe/io/csv.py [hunk 7]
```diff
@@ -660,6 +665,8 @@ def to_csv(
         Function accepting an integer (partition index) and producing a
         string to replace the asterisk in the given filename globstring.
         Should preserve the lexicographic order of partitions
+    single_file : bool, default False
+        Whether to save everything into a single CSV file.
```
Comment (mrocklin) [inline dask/dataframe/io/csv.py]: Thanks!

File: dask/dataframe/io/csv.py [hunk 8]
```diff
@@ -688,7 +695,8 @@ def to_csv(
     nanRep : None
         deprecated, use na_rep
     mode : str
-        Python write mode, default 'w'
+        Python write mode, default 'w'. Note that not all filesystems
+        support the append mode.
```
Comment (mrocklin) [inline dask/dataframe/io/csv.py]: Is this only important in the single file mode?  If so, I recommend that we put this note under `single_file`.  This is unlikely to be relevant to users in the many csv files case, which I expect to be more common.

File: dask/dataframe/io/csv.py [hunk 9]
```diff
@@ -725,50 +733,47 @@ def to_csv(
...
-    else:
-        default_encoding = "utf-8"
-        mode = "wt"
-
-    encoding = kwargs.get("encoding", default_encoding)
-
-    files = open_files(
-        filename,
+    if single_file and name_function is not None:
+        raise ValueError("name_function is not supported under the single file mode")
```
Comment (mrocklin) [inline dask/dataframe/io/csv.py]: Thank you for the informative error message.

File: dask/dataframe/io/csv.py (hunk 8 above)
Comment (mrocklin) [reply dask/dataframe/io/csv.py]: Also, are there file systems where this is known to fail or provide inaccurate results?  For example what about eventually consistent cloud storage systems like S3/GCS?  Should we err quickly if the user tries to do something like the following? 

```
df.to_csv("s3://...", single_file=True)
```

Comment (hongzmsft) [reply dask/dataframe/io/csv.py]: I have moved the note to the single_file entry.

I actually do not know how to deal with the S3/GCS case, and I am not familiar with the consistency models of these systems. A list of incompatible file systems may be hard to curate and change from time to time. My expectation is to follow the spirit of Python and trust the users. I have added a note regarding the network storage systems. An appropriate warning might also be good. I feel an error might be too strict. After all, if a user have some weird settings in her or his custom S3 storage server (Minio, Apache Oxide, or maybe others) and does want the behaviour, it would be unwise for us to prevent that.

Let me know of your opinion.

**Comment by hongzmsft** [comment]
@mrocklin Thank you for your review. Let me know of any other critiques.

File: dask/dataframe/io/csv.py [hunk 10]
```diff
@@ -660,6 +664,12 @@ def to_csv(
         Function accepting an integer (partition index) and producing a
         string to replace the asterisk in the given filename globstring.
         Should preserve the lexicographic order of partitions
+    single_file : bool, default False
+        Whether to save everything into a single CSV file. Under the
+        single file mode, each partition is appended at the end of the
+        specified CSV file. Note that not all filesystems support the
+        append mode and thus the single file mode, especially on cloud
+        storage systems such as S3 or GCS.
```
Comment (TomAugspurger) [inline dask/dataframe/io/csv.py]: What happens in this case? An exception is raised? It'd be nice to note that.

File: dask/dataframe/io/csv.py [hunk 11]
```diff
@@ -725,50 +735,47 @@ def to_csv(
...
+            name_function=name_function,
+            num=df.npartitions,
+            **file_options
+        )
+        values = [to_csv_chunk(dfs[0], files[0], None, **kwargs)]
         if header_first_partition_only:
             kwargs["header"] = False
         values.extend(
-            [to_csv_chunk(d, f, **kwargs) for d, f in zip(dfs[1:], files[1:])]
+            [to_csv_chunk(d, f, None, **kwargs) for d, f in zip(dfs[1:], files[1:])]
```
Comment (TomAugspurger) [inline dask/dataframe/io/csv.py]: Can you use a keyword argument here for `None`? I'm not sure what it's doing.

File: dask/dataframe/io/csv.py [hunk 12]
```diff
@@ -725,50 +735,47 @@ def to_csv(
...
-
     dfs = df.to_delayed()
-
-    # If we only want headers in the first file, turn headers off after the
-    # first partition
-
-    values = [to_csv_chunk(dfs[0], files[0], **kwargs)]
-
-    if len(dfs) > 1:
+    if single_file:
```
Comment (TomAugspurger) [inline dask/dataframe/io/csv.py]: Do we need a check somewhere in here for `header_first_partition_only`

**Comment by hongzmsft** [comment]
@TomAugspurger Fixed. Thanks.

File: dask/dataframe/io/csv.py [hunk 13]
```diff
@@ -724,51 +742,62 @@ def to_csv(
...
-        filename,
+    Raises
+    ------
+    ValueError
+        If `header_first_partition_only` is set to `False` or
+        `name_function` is specified when `single_file` is `True`.
+    """
+    if single_file and name_function is not None:
+        raise ValueError("name_function is not supported under the single file mode")
+    if header_first_partition_only is None:
```
Comment (TomAugspurger) [inline dask/dataframe/io/csv.py]: Thanks, this looks good. Could you add tests that hit the new behavior (I think just the `elif not on L756`).

**Comment by hongzmsft** [comment]
@TomAugspurger Fixed. Thanks.

File: dask/dataframe/io/tests/test_csv.py [hunk 14]
```diff
@@ -1287,6 +1287,61 @@ def test_to_csv_multiple_files_cornercases():
...
+
+        with tmpdir() as dn:
+            fn = os.path.join(dn, "test.csv")
+            r = a.to_csv(fn, index=False, compute=False, single_file=True)
+            dask.compute(r, scheduler="sync")
+            result = dd.read_csv(fn).compute().reset_index(drop=True)
+            assert_eq(result, df)
+
+
+@pytest.mark.xfail(
```
Comment (TomAugspurger) [inline dask/dataframe/io/tests/test_csv.py]: In my experience, these tests are written using `pytest.raises`, rather than an `xfail`.

```python
a = dd.from_pandas(df, 1)
with tmpdir() as dn:
    with pytest.raises(ValueError, match="name_function is not supported"):
        a.to_csv(...)
```

**Comment by hongzmsft** [comment]
@TomAugspurger Sorry, I was clearly not in my top form. Fixed. Thanks.

**Comment by TomAugspurger** [comment]
Thanks!
